plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeinfocache
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	clientcache "k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+)
+
+// New returns a NodeInfoCache implementation.
+// It automatically starts a go routine that manages expiration of assumed pods.
+// "ttl" is how long the assumed pod will get expired.
+// "period" is how long the background goroutine should wait before cleaning up expired pods periodically.
+// "stop" is the channel that signals stopping and we would close background goroutines.
+func New(ttl, period time.Duration, stop chan struct{}) schedulercache.NodeInfoCache {
+	cache := newNodeInfoCache(ttl, period, stop)
+	cache.run()
+	return cache
+}
+
+// mustGetPodKey returns the string key of a pod.
+// A pod is ensured to have accessor. We don't want to check the error everytime.
+// TODO: We should consider adding a Key() method to api.Pod
+func mustGetPodKey(pod *api.Pod) string {
+	key, err := clientcache.MetaNamespaceKeyFunc(pod)
+	if err != nil {
+		panic("api.Pod should have key func: " + err.Error())
+	}
+	return key
+}
+
+type nodeInfoCache struct {
+	stop chan struct{}
+
+	// This mutex guards all fields within this cache struct
+	mu          sync.Mutex
+	ttl         time.Duration
```

Comment:
This is related to assumed pods' logic. See https://github.com/kubernetes/kubernetes/blob/15b30b8b099377b1ddd6868aa13a3141b9e2ac28/plugin/pkg/scheduler/modeler.go#L108

In my understanding, this prevents failed or slow detection of unscheduled pods.
See the state machine drawn in this PR, assumed pods need to be confirmed, otherwise it would expire.